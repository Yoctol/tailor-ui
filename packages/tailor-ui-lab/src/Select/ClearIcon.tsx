import React, { FC } from 'react';
import styled from 'styled-components';
import { MdClose } from 'react-icons/md';

import { Icon } from 'tailor-ui';

import { getDataTestId } from './utils';

interface ClearIconProps {
  clearSelection: () => void;
  'data-testid'?: string;
}

const ClearIconWrapper = styled.div`
  display: inline-flex;
  flex: none;

  &:hover svg {
    fill: ${p => p.theme.colors.gray700} !important;
  }
`;

const ClearIcon: FC<ClearIconProps> = ({ clearSelection, ...props }) => (
  <ClearIconWrapper>
    <Icon
      type={MdClose}
      fill="gray400"
      size="20"
      cursor="pointer"
      onClick={e => {
        e.stopPropagation();
        clearSelection();
      }}
      {...getDataTestId(props, 'clear-icon')}
    />
  </ClearIconWrapper>
);

export default ClearIcon;
