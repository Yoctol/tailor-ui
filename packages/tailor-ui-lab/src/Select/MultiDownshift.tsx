import Downshift, {
  ControllerStateAndHelpers,
  DownshiftInterface,
} from 'downshift';
import React, { Component, ReactNode } from 'react';

import { itemToString } from './utils';

type Option =
  | {
      label: string;
      value: string;
    }
  | string
  | number;

type MultiDownshiftProps = DownshiftInterface<Option> & {
  selectedItem: Option[];
  onSelect?: (
    options: Option[],
    stateAndHelpers: ControllerStateAndHelpers<Option>
  ) => void;
  onChange?: (
    options: Option[],
    stateAndHelpers: ControllerStateAndHelpers<Option>
  ) => void;
  children: (options: ControllerStateAndHelpers<Option>) => ReactNode;
};

interface MultiDownshiftState {
  selectedItems: Option[];
}

class MultiDownshift extends Component<
  MultiDownshiftProps,
  MultiDownshiftState
> {
  state: MultiDownshiftState = {
    selectedItems: this.props.selectedItem || [],
  };

  stateReducer = (state: any, changes: any) => {
    switch (changes.type) {
      case Downshift.stateChangeTypes.blurButton:
        return {};
      case Downshift.stateChangeTypes.keyDownEnter:
      case Downshift.stateChangeTypes.mouseUp:
      case Downshift.stateChangeTypes.clickItem:
        return {
          ...changes,
          highlightedIndex: state.highlightedIndex,
          isOpen: true,
          inputValue: '',
        };
      default:
        return changes;
    }
  };

  handleSelection = (selectedItem: Option, downshift: any) => {
    const callOnChange = () => {
      const { onSelect, onChange } = this.props;
      const { selectedItems } = this.state;
      if (onSelect) {
        onSelect(selectedItems, this.getStateAndHelpers(downshift));
      }
      if (onChange) {
        onChange(selectedItems, this.getStateAndHelpers(downshift));
      }
    };

    if (
      this.state.selectedItems
        .map(itemToString)
        .includes(itemToString(selectedItem))
    ) {
      this.removeItem(selectedItem, callOnChange);
    } else {
      this.addSelectedItem(selectedItem, callOnChange);
    }
  };

  removeItem = (item: Option, cb?: () => void) => {
    this.setState(({ selectedItems }) => {
      return {
        selectedItems: selectedItems.filter(
          i => itemToString(i) !== itemToString(item)
        ),
      };
    }, cb);
  };

  addSelectedItem(item: Option, cb: () => void) {
    this.setState(
      ({ selectedItems }) => ({
        selectedItems: [...selectedItems, item],
      }),
      cb
    );
  }

  getRemoveButtonProps = ({
    onClick,
    item,
    ...props
  }: {
    onClick?: (e: any) => void;
    item: Option;
  }) => {
    return {
      onClick: (e: any) => {
        // TODO: use something like downshift's composeEventHandlers utility instead
        if (onClick) {
          onClick(e);
        }
        e.stopPropagation();
        this.removeItem(item);
      },
      ...props,
    };
  };

  getStateAndHelpers(downshift: ControllerStateAndHelpers<any>) {
    const { selectedItems } = this.state;
    const { getRemoveButtonProps, removeItem } = this;
    return {
      getRemoveButtonProps,
      removeItem,
      selectedItems,
      ...downshift,
    };
  }

  render() {
    const { children, ...props } = this.props;
    // TODO: compose together props (rather than overwriting them) like downshift does
    return (
      <Downshift
        {...props}
        stateReducer={this.stateReducer}
        onChange={this.handleSelection}
        selectedItem={null}
      >
        {downshift => children(this.getStateAndHelpers(downshift))}
      </Downshift>
    );
  }
}

export default MultiDownshift;
