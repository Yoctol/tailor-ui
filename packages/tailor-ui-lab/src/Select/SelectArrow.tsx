import React, { FC } from 'react';
import { MdKeyboardArrowUp } from 'react-icons/md';
import { animated, config, useSpring } from 'react-spring';

import { Icon } from 'tailor-ui';

const SelectArrow: FC<{ on: boolean }> = ({ on }) => {
  const { deg } = useSpring({
    deg: on ? 0 : 180,
    config: config.stiff,
  });

  return (
    <animated.i
      style={{
        transform: deg.interpolate(d => `rotate(${d}deg)`),
        flex: 'none',
        pointerEvents: 'none',
        display: 'inline-flex',
      }}
    >
      <Icon type={MdKeyboardArrowUp} fill="gray400" />
    </animated.i>
  );
};

export default SelectArrow;
