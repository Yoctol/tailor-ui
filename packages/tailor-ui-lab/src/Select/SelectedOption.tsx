import React, { FunctionComponent } from 'react';
import { MdClose } from 'react-icons/md';

import { Icon } from 'tailor-ui';

import { StyledSelectedOption } from './styles';
import { getDataTestId } from './utils';

export interface SelectedOption {
  index: number;
  'data-testid'?: string;
}

const SelectedOption: FunctionComponent<SelectedOption> = ({
  children,
  index,
  ...props
}) => {
  return (
    <StyledSelectedOption {...getDataTestId(props, `selected-option-${index}`)}>
      {children}
      <Icon
        type={MdClose}
        size="16"
        cursor="pointer"
        ml="1"
        {...props}
        {...getDataTestId(props, `selected-option-${index}-clear-icon`)}
      />
    </StyledSelectedOption>
  );
};

export default SelectedOption;
