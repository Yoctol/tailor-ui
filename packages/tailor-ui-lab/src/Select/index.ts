import { Option as _Option } from './SelectOptions';

export * from './Select';
export type Option = _Option;
