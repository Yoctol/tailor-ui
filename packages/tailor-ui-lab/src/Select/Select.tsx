import Downshift from 'downshift';
import React, { FunctionComponent, ReactNode, useRef, useState } from 'react';

import { Flex, Popover, Position, useFormField } from 'tailor-ui';

import ClearIcon from './ClearIcon';
import MultiDownshift from './MultiDownshift';
import SelectArrow from './SelectArrow';
import SelectInput from './SelectInput';
import SelectOptions, { ObjectOption, Option } from './SelectOptions';
import SelectedOption from './SelectedOption';
import { Loading, SelectWrapper, StyledSelect } from './styles';
import { getDataTestId, itemToString } from './utils';

interface SelectProps {
  id?: string;
  name?: string;
  className?: string;
  width?: string | number;
  size?: 'sm' | 'md' | 'lg';
  creatable?: boolean;
  clearable?: boolean;
  disabled?: boolean;
  loading?: boolean;
  multiple?: boolean;
  searchable?: boolean;
  options: Option[];
  value?: Option;
  defaultValue?: Option;
  placeholder?: string;
  menu?: ReactNode;
  itemSize?: number;
  optionsMaxHeight?: number;
  onChange?: (option: Option) => void;
  noOptionsMessage?: () => ReactNode;
  formatCreateLabel?: (labelInfo: {
    value: string;
    active: boolean;
    hovered: boolean;
  }) => ReactNode;
  isValidNewOption?: (value: string) => boolean;
  onCreateOption?: (value: string) => void;
  'data-testid'?: string;
}

const Select: FunctionComponent<SelectProps> = ({
  id,
  width = 240,
  size = 'md',
  creatable = false,
  clearable = false,
  disabled = false,
  loading = false,
  multiple = false,
  searchable = false,
  options,
  value,
  defaultValue,
  placeholder = '',
  menu,
  itemSize = 36,
  optionsMaxHeight = 180,
  onChange,
  noOptionsMessage,
  formatCreateLabel,
  isValidNewOption,
  onCreateOption,
  ...props
}) => {
  const [invalid, labelId, setValue] = useFormField({
    id,
    value,
    defaultValue,
  });
  const inputRef = useRef<HTMLInputElement>(null);
  const [visible, setVisible] = useState(false);
  const [inputValue, setInputValue] = useState('');

  const handleChange = (selection: Option) => {
    if (inputRef.current && !multiple) {
      inputRef.current.blur();
    }

    const isCreate =
      selection && (selection as ObjectOption).label === 'CREATE_OPTION';

    if (isCreate && onCreateOption) {
      onCreateOption((selection as ObjectOption).value);
    }

    if (!isCreate && onChange) {
      onChange(selection);
      setValue(selection);
    }

    if (!multiple || isCreate) {
      setVisible(false);
    } else {
      setInputValue('');
    }
  };

  const defaultHighlightedIndex = multiple
    ? null
    : options.findIndex(
        option => itemToString(option) === itemToString(value || defaultValue)
      ) || null;

  const RenderComponent = multiple ? MultiDownshift : Downshift;

  return (
    <RenderComponent
      id={labelId}
      selectedItem={value}
      isOpen={visible}
      initialSelectedItem={defaultValue}
      defaultHighlightedIndex={defaultHighlightedIndex}
      stateReducer={(state, changes) => {
        // if the user is opening the menu, then let's make sure
        // that the highlighted index is set to the selected index
        if (
          typeof changes === 'object' &&
          changes.type === Downshift.stateChangeTypes.mouseUp
        ) {
          return {};
        }

        return changes;
      }}
      itemToString={itemToString}
      inputValue={inputValue}
      onInputValueChange={newInputValue => {
        if (!multiple && newInputValue !== 'CREATE_OPTION') {
          setInputValue(newInputValue);
        }
      }}
      onChange={handleChange}
      {...{
        scrollIntoView: () => {},
      } as any}
    >
      {({
        id: downshiftId,

        getRootProps,
        getMenuProps,
        getItemProps,
        getInputProps,
        highlightedIndex,
        selectedItem,
        clearSelection,

        selectedItems = [],
        getRemoveButtonProps,
        removeItem,
      }: any) => (
        <SelectWrapper
          {...getRootProps({
            style: { width },
          })}
        >
          <Popover
            visible={visible}
            position={Position.BOTTOM_LEFT}
            minWidth={width}
            onVisibleChange={newVisible => {
              if (disabled || loading) {
                return;
              }

              if (!newVisible && inputRef.current) {
                inputRef.current.blur();
              }

              setVisible(newVisible);
            }}
            p="0"
            content={
              <SelectOptions
                visible={visible}
                creatable={creatable}
                searchable={searchable}
                multiple={multiple}
                itemSize={itemSize}
                optionsMaxHeight={optionsMaxHeight}
                getMenuProps={getMenuProps}
                getItemProps={getItemProps}
                options={options}
                inputValue={inputValue}
                menu={menu}
                highlightedIndex={highlightedIndex}
                selectedItem={selectedItem}
                selectedItems={selectedItems}
                noOptionsMessage={noOptionsMessage}
                formatCreateLabel={formatCreateLabel}
                isValidNewOption={isValidNewOption}
                data-testid={props['data-testid']}
              />
            }
          >
            <StyledSelect
              id={downshiftId}
              {...getDataTestId(props)}
              invalid={invalid}
              size={size}
              focused={visible}
              disabled={disabled || loading}
              onClick={() => {
                if (!visible) {
                  if (searchable || creatable || multiple) {
                    setInputValue('');
                  }
                  if (inputRef.current) {
                    inputRef.current.focus();
                  }
                }
              }}
            >
              <Flex
                flex="auto"
                flexWrap="wrap"
                alignItems="center"
                overflow="hidden"
              >
                {multiple &&
                  selectedItems.map((item: Option) => (
                    <SelectedOption
                      key={itemToString(item)}
                      {...getRemoveButtonProps({
                        item,
                        onClick: () => {
                          if (inputRef.current) {
                            inputRef.current.focus();
                          }
                        },
                      })}
                    >
                      <span>{itemToString(item)}</span>
                    </SelectedOption>
                  ))}
                <SelectInput
                  ref={inputRef}
                  visible={visible}
                  searchable={searchable}
                  creatable={creatable}
                  multiple={multiple}
                  inputValue={inputValue}
                  placeholder={placeholder}
                  selectedItem={selectedItem}
                  selectedItems={selectedItems}
                  removeItem={removeItem}
                  onChange={event => setInputValue(event.currentTarget.value)}
                  getInputProps={getInputProps}
                  data-testid={props['data-testid']}
                />
              </Flex>
              {clearable && selectedItem && (
                <ClearIcon clearSelection={clearSelection} />
              )}
              {loading ? (
                <Loading title="loading" />
              ) : (
                <SelectArrow on={visible} />
              )}
            </StyledSelect>
          </Popover>
        </SelectWrapper>
      )}
    </RenderComponent>
  );
};

export { Select };
