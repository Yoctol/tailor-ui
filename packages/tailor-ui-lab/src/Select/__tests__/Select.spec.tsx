import React, { useState } from 'react';

import { fireEvent, render, wait, waitForElement } from 'test/test-utils';

import { Select } from '../Select';

const DEFAULT_OPTIONS = [
  { label: 'Banana', value: 'banana' },
  { label: 'Orange', value: 'orange' },
  { label: 'Apple', value: 'apple' },
  { label: 'Mango', value: 'mango' },
];

const DEFAULT_SELECT_OPTION = { label: 'Banana', value: 'banana' };

const ControlSelect = (props: any) => {
  const [value, setValue] = useState<any>(DEFAULT_SELECT_OPTION);

  return (
    <Select
      value={value}
      onChange={newValue => setValue(newValue)}
      options={DEFAULT_OPTIONS}
      {...props}
    />
  );
};

describe('Select', () => {
  it('should render correctly', () => {
    const { baseElement } = render(<Select options={DEFAULT_OPTIONS} />);

    expect(baseElement).toMatchSnapshot();
  });

  it('should not open menu when disabled', () => {
    const { getByTestId, queryByTestId } = render(
      <ControlSelect data-testid="select" disabled />
    );

    const select = getByTestId('select');
    fireEvent.click(select);

    expect(queryByTestId('select-menu')).not.toBeInTheDocument();
  });

  it('should not open menu when loading', () => {
    const { getByTestId, queryByTestId } = render(
      <ControlSelect data-testid="select" loading />
    );

    const select = getByTestId('select');
    fireEvent.click(select);

    expect(queryByTestId('select-menu')).not.toBeInTheDocument();
  });

  it('should change value correctly when click option', async () => {
    const { getByTestId } = render(<ControlSelect data-testid="select" />);

    const input = getByTestId('select-input') as HTMLInputElement;
    expect(input.value).toBe('Banana');

    const select = getByTestId('select');
    fireEvent.click(select);

    await waitForElement(() => getByTestId('select-menu'));

    const orange = getByTestId('select-item-1');
    fireEvent.click(orange);

    expect(input.value).toBe('Orange');
  });

  describe('searchable', () => {
    it('should search correctly when change select input', async () => {
      const { getByTestId, queryByTestId, queryByText } = render(
        <ControlSelect searchable data-testid="select" />
      );

      const input = getByTestId('select-input') as HTMLInputElement;
      expect(input.value).toBe('Banana');

      const select = getByTestId('select');
      fireEvent.click(select);

      await waitForElement(() => getByTestId('select-menu'));

      expect(input.value).toBe('');
      expect(input.placeholder).toBe('Banana');

      fireEvent.change(input, {
        target: {
          value: 'Orange',
        },
      });

      const orange = getByTestId('select-item-0');
      fireEvent.click(orange);

      expect(getByTestId('select-item-0')).toHaveTextContent('Orange');
      expect(queryByTestId('select-item-1')).not.toBeInTheDocument();
      expect(queryByText('Banana')).not.toBeInTheDocument();
    });
  });

  describe('creatable', () => {
    it('should create successfully when pass creatable', async () => {
      const CreatableSelect = () => {
        const [loading, setLoading] = useState(false);
        const [value, setValue] = useState<any>(DEFAULT_SELECT_OPTION);
        const [options, setOptions] = useState(DEFAULT_OPTIONS);

        return (
          <Select
            creatable
            data-testid="select"
            loading={loading}
            value={value}
            onChange={newValue => setValue(newValue)}
            options={options}
            isValidNewOption={name =>
              !options.map(option => option.value).includes(name) &&
              name.trim() !== ''
            }
            onCreateOption={name => {
              const newOption = { label: name, value: name };
              setLoading(true);
              setTimeout(() => {
                setOptions([...options, newOption]);
                setValue(newOption);
                setLoading(false);
              }, 1000);
            }}
          />
        );
      };

      jest.useFakeTimers();

      const { getByTestId, queryByTitle, getByText, getByTitle } = render(
        <CreatableSelect />
      );

      const input = getByTestId('select-input') as HTMLInputElement;
      expect(input.value).toBe('Banana');

      const select = getByTestId('select');
      fireEvent.click(select);

      await waitForElement(() => getByTestId('select-menu'));

      fireEvent.change(input, {
        target: {
          value: 'XXXXXX',
        },
      });

      const newOption = getByText('Create new option: XXXXXX');
      expect(newOption).toBeInTheDocument();
      fireEvent.click(newOption);

      expect(getByTitle('loading')).toBeInTheDocument();

      jest.runAllTimers();

      expect(queryByTitle('loading')).not.toBeInTheDocument();
      expect(input.value).toBe('XXXXXX');

      jest.useRealTimers();
    });

    it('should not creatable when option is not valid', async () => {
      const CreatableSelect = () => {
        const [value, setValue] = useState<any>(DEFAULT_SELECT_OPTION);

        return (
          <Select
            creatable
            data-testid="select"
            value={value}
            onChange={newValue => setValue(newValue)}
            options={DEFAULT_OPTIONS}
            isValidNewOption={name => name === 'NOT_VALID'}
          />
        );
      };

      const { getByTestId, queryByText } = render(<CreatableSelect />);

      const input = getByTestId('select-input') as HTMLInputElement;
      expect(input.value).toBe('Banana');

      const select = getByTestId('select');
      fireEvent.click(select);

      await waitForElement(() => getByTestId('select-menu'));

      fireEvent.change(input, {
        target: {
          value: 'NOT_VALID',
        },
      });

      const newOption = queryByText('Create new option: NOT_VALID');
      expect(newOption).toBeInTheDocument();
    });
  });

  describe('clearable', () => {
    it('should clear value when click clear icon', async () => {
      const ClearableSelect = () => {
        const [value, setValue] = useState<any>(DEFAULT_SELECT_OPTION);

        return (
          <Select
            clearable
            data-testid="select"
            value={value}
            onChange={newValue => setValue(newValue)}
            options={DEFAULT_OPTIONS}
          />
        );
      };

      const { getByTestId } = render(<ClearableSelect />);

      const input = getByTestId('select-input') as HTMLInputElement;
      expect(input.value).toBe('Banana');

      const clearIcon = getByTestId('select-clear-icon');
      fireEvent.click(clearIcon);

      expect(input.value).toBe('');
    });
  });

  describe('multiple', () => {
    const MultipleSelect = () => {
      const [loading, setLoading] = useState(false);
      const [value, setValue] = useState([
        { label: 'Banana', value: 'banana' },
        { label: 'Orange', value: 'orange' },
      ]);
      const [options, setOptions] = useState([
        { label: 'Banana', value: 'banana' },
        { label: 'Orange', value: 'orange' },
        { label: 'Apple', value: 'apple' },
        { label: 'Mango', value: 'mango' },
      ]);

      return (
        <Select
          creatable
          data-testid="select"
          loading={loading}
          width="360px"
          multiple
          value={value}
          onChange={newValue => setValue(newValue as any)}
          options={options}
          isValidNewOption={name =>
            !options.map(option => option.label).includes(name) &&
            name.trim() !== ''
          }
          onCreateOption={name => {
            const newOption = { label: name, value: name };
            setLoading(true);
            setTimeout(() => {
              setOptions([...options, newOption]);
              setValue(prevValue => [...prevValue, newOption]);
              setLoading(false);
            }, 1000);
          }}
        />
      );
    };

    it('should display multi option', () => {
      const { getByText, baseElement } = render(<MultipleSelect />);

      expect(baseElement).toMatchSnapshot();

      expect(getByText('Banana')).toBeInTheDocument();
      expect(getByText('Orange')).toBeInTheDocument();
    });

    it('should add selected option when click other option', async () => {
      const { getByTestId, getByText, queryByTestId } = render(
        <MultipleSelect />
      );

      const select = getByTestId('select');
      fireEvent.click(select);

      await waitForElement(() => getByTestId('select-menu'));

      fireEvent.click(getByText('Apple'));

      expect(getByTestId('select-menu')).toBeInTheDocument();

      fireEvent.click(getByText('Mango'));

      fireEvent.click(select);

      await wait(() =>
        expect(queryByTestId('select-menu')).not.toBeInTheDocument()
      );
      expect(getByText('Apple')).toBeInTheDocument();
      expect(getByText('Mango')).toBeInTheDocument();
    });

    it('should remove selected option when click clear ', async () => {
      const { getByTestId, queryByText } = render(<MultipleSelect />);

      fireEvent.click(getByTestId('select-selected-option-1-clear-icon'));
      fireEvent.click(getByTestId('select-selected-option-0-clear-icon'));

      expect(queryByText('Banana')).not.toBeInTheDocument();
      expect(queryByText('Orange')).not.toBeInTheDocument();
    });

    it('should remove last selected option when press backspace ', async () => {
      const { getByTestId, queryByText, getByText, queryByTestId } = render(
        <MultipleSelect />
      );

      const select = getByTestId('select');
      fireEvent.click(select);

      const input = getByTestId('select-input') as HTMLInputElement;
      fireEvent.keyDown(input, {
        key: 'Backspace',
        currentTarget: {
          value: '',
        },
      });

      fireEvent.click(select);

      await wait(() =>
        expect(queryByTestId('select-menu')).not.toBeInTheDocument()
      );

      expect(getByText('Banana')).toBeInTheDocument();
      expect(queryByText('Orange')).not.toBeInTheDocument();
    });
  });
});
