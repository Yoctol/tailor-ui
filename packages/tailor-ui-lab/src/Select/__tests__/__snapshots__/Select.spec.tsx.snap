// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select multiple should display multi option 1`] = `
<body>
  <div
    portal-container=""
  />
  <div
    aria-live="polite"
    aria-relevant="additions text"
    id="a11y-status-message"
    role="status"
    style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px;"
  />
  .c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  overflow: hidden;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  cursor: default;
}

.c7 svg {
  fill: #8b8b8b;
  vertical-align: middle;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c7.c7.c7 svg {
  fill: #bfbfbf;
  width: 24px;
  height: 24px;
}

.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  cursor: pointer;
  margin-left: 0.25rem;
}

.c5 svg {
  fill: #8b8b8b;
  vertical-align: middle;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c5.c5.c5 svg {
  width: 16px;
  height: 16px;
}

.c6 {
  visibility: hidden;
  position: absolute;
  pointer-events: none;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  cursor: pointer;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-width: 100%;
  height: 28px;
  margin: 2px 8px 2px 0;
  padding: 0 8px;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #6B88c5;
  background-color: #f8f8fc;
  color: #6B88c5;
  font-size: 0.875rem;
}

.c4 > span {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.c4 svg {
  fill: #d5d8e8;
}

.c4:hover svg {
  fill: #6B88c5;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #bfbfbf;
  outline: none;
  background-color: #ffffff;
  color: #373737;
  line-height: 1.5;
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  min-height: 2.25rem;
  padding: 4px 0.75rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c1:hover {
  border-color: #3c5ad0;
}

.c1 input {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: 100%;
  overflow: hidden;
  border: none;
  outline: none;
  background-color: transparent;
  text-overflow: ellipsis;
  pointer-events: none;
}

.c1 input::-webkit-input-placeholder {
  color: #bfbfbf;
}

.c1 input::-moz-placeholder {
  color: #bfbfbf;
}

.c1 input:-ms-input-placeholder {
  color: #bfbfbf;
}

.c1 input::placeholder {
  color: #bfbfbf;
}

.c1 input {
  font-size: 1rem;
}

.c8 .c3 {
  border-color: #bfbfbf;
  background-color: #f6f6f6;
  color: #bfbfbf;
  pointer-events: none;
}

<div>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="downshift-8-label"
      class="c0"
      role="combobox"
      style="width: 360px;"
    >
      <div
        class="c1"
        data-testid="select"
        id="downshift-8"
      >
        <div
          class="c2"
          overflow="hidden"
        >
          <div
            class="c3 c4"
            data-testid="select-selected-option-0"
          >
            <span>
              Banana
            </span>
            <i
              class="c5"
              cursor="pointer"
              data-testid="select-selected-option-0-clear-icon"
              size="16"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </i>
          </div>
          <div
            class="c3 c4"
            data-testid="select-selected-option-1"
          >
            <span>
              Orange
            </span>
            <i
              class="c5"
              cursor="pointer"
              data-testid="select-selected-option-1-clear-icon"
              size="16"
            >
              <svg
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </i>
          </div>
          <input
            aria-autocomplete="list"
            aria-labelledby="downshift-8-label"
            autocomplete="off"
            data-testid="select-input"
            id="downshift-8-input"
            placeholder=""
            readonly=""
            style="width: 18px; box-sizing: content-box;"
            value=""
          />
          <span
            class="c6"
            style="font-size: 16px;"
          />
        </div>
        <i
          style="transform: rotate(180deg); pointer-events: none; display: inline-flex;"
        >
          <i
            class="c7"
            cursor="default"
            fill="gray400"
            size="24"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
              />
            </svg>
          </i>
        </i>
      </div>
    </div>
  </div>
</body>
`;

exports[`Select should render correctly 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  overflow: hidden;
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1;
  cursor: default;
}

.c4 svg {
  fill: #8b8b8b;
  vertical-align: middle;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c4.c4.c4 svg {
  fill: #bfbfbf;
  width: 24px;
  height: 24px;
}

.c3 {
  visibility: hidden;
  position: absolute;
  pointer-events: none;
}

.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  cursor: pointer;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  max-width: 100%;
  border: 1px solid;
  border-radius: 0.25rem;
  border-color: #bfbfbf;
  outline: none;
  background-color: #ffffff;
  color: #373737;
  line-height: 1.5;
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  min-height: 2.25rem;
  padding: 4px 0.75rem;
  -webkit-transition: all 200ms ease 0s;
  transition: all 200ms ease 0s;
}

.c1:hover {
  border-color: #3c5ad0;
}

.c1 input {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: 100%;
  overflow: hidden;
  border: none;
  outline: none;
  background-color: transparent;
  text-overflow: ellipsis;
  pointer-events: none;
}

.c1 input::-webkit-input-placeholder {
  color: #bfbfbf;
}

.c1 input::-moz-placeholder {
  color: #bfbfbf;
}

.c1 input:-ms-input-placeholder {
  color: #bfbfbf;
}

.c1 input::placeholder {
  color: #bfbfbf;
}

.c1 input {
  font-size: 1rem;
}

<body>
  <div>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="downshift-0-label"
      class="c0"
      role="combobox"
      style="width: 240px;"
    >
      <div
        class="c1"
        id="downshift-0"
      >
        <div
          class="c2"
          overflow="hidden"
        >
          <input
            aria-autocomplete="list"
            aria-labelledby="downshift-0-label"
            autocomplete="off"
            id="downshift-0-input"
            placeholder=""
            readonly=""
            style="width: 18px; box-sizing: content-box;"
            value=""
          />
          <span
            class="c3"
            style="font-size: 16px;"
          />
        </div>
        <i
          style="transform: rotate(180deg); pointer-events: none; display: inline-flex;"
        >
          <i
            class="c4"
            cursor="default"
            fill="gray400"
            size="24"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M7.41 15.41L12 10.83l4.59 4.58L18 14l-6-6-6 6z"
              />
            </svg>
          </i>
        </i>
      </div>
    </div>
  </div>
</body>
`;
